In [1]:
import pandas as pd

In [2]:
data=pd.read_csv('/Users/kavisha/Documents/Springboard/listingsNY.csv', low_memory=False)

In [3]:
pd.options.display.max_columns = 100

In [4]:
data.head()

id                         listing_url       scrape_id last_scraped  \
0   2515   https://www.airbnb.com/rooms/2515  20181206022948   2018-12-06   
1  21456  https://www.airbnb.com/rooms/21456  20181206022948   2018-12-06   
2   2539   https://www.airbnb.com/rooms/2539  20181206022948   2018-12-06   
3   2595   https://www.airbnb.com/rooms/2595  20181206022948   2018-12-06   
4  21644  https://www.airbnb.com/rooms/21644  20181206022948   2018-12-06   

                                 name  \
0    Stay at Chez Chic budget room #1   
1  Light-filled classic Central Park    
2  Clean & quiet apt home by the park   
3               Skylit Midtown Castle   
4           Upper Manhattan, New York   

                                             summary  \
0  Step into our artistic spacious apartment and ...   
1  An adorable, classic, clean, light-filled one-...   
2           Renovated apt home in elevator building.   
3  Find your romantic getaway to this beautiful, ...   
4  A great space in a beautiful neighborhood- min...   

                                               space  \
0  -PLEASE BOOK DIRECTLY. NO NEED TO SEND A REQUE...   
1  An adorable, classic, clean, light-filled one-...   
2  Spacious, renovated, and clean apt home, one b...   
3  - Spacious (500+ft²), immaculate and nicely fu...   
4  Nice room in a spacious pre-war apartment in u...   

                                         description experiences_offered  \
0  Step into our artistic spacious apartment and ...                none   
1  An adorable, classic, clean, light-filled one-...                none   
2  Renovated apt home in elevator building. Spaci...                none   
3  Find your romantic getaway to this beautiful, ...                none   
4  A great space in a beautiful neighborhood- min...                none   

                               neighborhood_overview  \
0                                                NaN   
1  Diverse. Great coffee shops and restaurants, n...   
2    Close to Prospect Park and Historic Ditmas Park   
3  Centrally located in the heart of Manhattan ju...   
4  I love that the neighborhood is safe to walk a...   

                                               notes  \
0  Please no cooking at night but you can warm up...   
1                                                NaN   
2  If this room is unavailable on your desired da...   
3                                                NaN   
4                                                NaN   

                                             transit  \
0  Subway 2.3.B.C. at 110th street around the cor...   
1  Yes!  B, C, and 1 trains are within 5 minute w...   
2  Very close to F and G trains and Express bus i...   
3  Apartment is located on 37th Street between 5t...   
4  You have access to local trains and express li...   

                                              access  \
0  Guests will have their PRIVATE BATHROOM (NOTE:...   
1  Guests have access to the whole apartment:  wi...   
2                                                NaN   
3  Guests have full access to the kitchen, bathro...   
4  guest will have access to the common living qu...   

                                         interaction  \
0  We will have a list of Harlem restaurants and ...   
1  I interact with guests upon arrival & departur...   
2                                                NaN   
3  I am a Sound Therapy Practitioner and Kundalin...   
4  I am available to serve as a resource to my gu...   

                                         house_rules  thumbnail_url  \
0  no-smoking/please take off your shoes: cleanin...            NaN   
1  Sorry, no pets and no smoking, please.   No pa...            NaN   
2  -The security and comfort of all our guests is...            NaN   
3  Make yourself at home, respect the space and t...            NaN   
4         Guest should be respectful of the space...            NaN   

   medium_url                                        

In [5]:
data.shape

(49056, 96)

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49056 entries, 0 to 49055
Data columns (total 96 columns):
id                                  49056 non-null int64
listing_url                         49056 non-null object
scrape_id                           49056 non-null int64
last_scraped                        49056 non-null object
name                                49039 non-null object
summary                             47140 non-null object
space                               34354 non-null object
description                         48522 non-null object
experiences_offered                 49056 non-null object
neighborhood_overview               29889 non-null object
notes                               20013 non-null object
transit                             30935 non-null object
access                              28549 non-null object
interaction                         27835 non-null object
house_rules                         29660 non-null object
thumbnail_url          

There are a total of 96 columns in the dataset. First step is to remove the unnecessary and redundant data columns. The columns where >50% of the values are null can be deleted. Total rows (listings) are 49056, and half of that is 24528. Accordingly, all columns with more than 24528 null values will be removed.

In [7]:
#Selecting only the columns with more than 50% non-null values and assigning them back to the dataframe.
data=data.loc[data.count(1) > data.shape[1]/2, data.count(0) > data.shape[0]/2]

In [8]:
data.shape

(49056, 86)

This step eliminated 10 columns with excessive number of null values. 

In [9]:
data.head()

id                         listing_url       scrape_id last_scraped  \
0   2515   https://www.airbnb.com/rooms/2515  20181206022948   2018-12-06   
1  21456  https://www.airbnb.com/rooms/21456  20181206022948   2018-12-06   
2   2539   https://www.airbnb.com/rooms/2539  20181206022948   2018-12-06   
3   2595   https://www.airbnb.com/rooms/2595  20181206022948   2018-12-06   
4  21644  https://www.airbnb.com/rooms/21644  20181206022948   2018-12-06   

                                 name  \
0    Stay at Chez Chic budget room #1   
1  Light-filled classic Central Park    
2  Clean & quiet apt home by the park   
3               Skylit Midtown Castle   
4           Upper Manhattan, New York   

                                             summary  \
0  Step into our artistic spacious apartment and ...   
1  An adorable, classic, clean, light-filled one-...   
2           Renovated apt home in elevator building.   
3  Find your romantic getaway to this beautiful, ...   
4  A great space in a beautiful neighborhood- min...   

                                               space  \
0  -PLEASE BOOK DIRECTLY. NO NEED TO SEND A REQUE...   
1  An adorable, classic, clean, light-filled one-...   
2  Spacious, renovated, and clean apt home, one b...   
3  - Spacious (500+ft²), immaculate and nicely fu...   
4  Nice room in a spacious pre-war apartment in u...   

                                         description experiences_offered  \
0  Step into our artistic spacious apartment and ...                none   
1  An adorable, classic, clean, light-filled one-...                none   
2  Renovated apt home in elevator building. Spaci...                none   
3  Find your romantic getaway to this beautiful, ...                none   
4  A great space in a beautiful neighborhood- min...                none   

                               neighborhood_overview  \
0                                                NaN   
1  Diverse. Great coffee shops and restaurants, n...   
2    Close to Prospect Park and Historic Ditmas Park   
3  Centrally located in the heart of Manhattan ju...   
4  I love that the neighborhood is safe to walk a...   

                                             transit  \
0  Subway 2.3.B.C. at 110th street around the cor...   
1  Yes!  B, C, and 1 trains are within 5 minute w...   
2  Very close to F and G trains and Express bus i...   
3  Apartment is located on 37th Street between 5t...   
4  You have access to local trains and express li...   

                                              access  \
0  Guests will have their PRIVATE BATHROOM (NOTE:...   
1  Guests have access to the whole apartment:  wi...   
2                                                NaN   
3  Guests have full access to the kitchen, bathro...   
4  guest will have access to the common living qu...   

                                         interaction  \
0  We will have a list of Harlem restaurants and ...   
1  I interact with guests upon arrival & departur...   
2                                                NaN   
3  I am a Sound Therapy Practitioner and Kundalin...   
4  I am available to serve as a resource to my gu...   

                                         house_rules  \
0  no-smoking/please take off your shoes: cleanin...   
1  Sorry, no pets and no smoking, please.   No pa...   
2  -The security and comfort of all our guests is...   
3  Make yourself at home, respect the space and t...   
4         Guest should be respectful of the space...   

                                         picture_url  host_id  \
0  https://a0.muscache.com/im/pictures/d0489e42-4...     2758   
1  https://a0.muscache.com/im/pictures/111808/a94...    42032   
2  https://a0.muscache.com/im/pictures/3949d073-a...     2787   
3  https://a0.muscache.com/im/pictures/f028bdf9-e...     2845   
4  https://a0.muscache.com/im/pictures/43197335/5...    82685   

                                  host_url host_name  host_since  \
0   https://www.airbn

# Data dictionary for the remaining columns to help determine which ones would be useful features:

> id - unique identifier number for the listing <br>
listing_url - url where the listing is found (contains the id) <br>
scrape_id - ??? <br>
last_scraped - ??? <br>
name - name of the property	<br>
summary	- summary of the property provided by the host<br>
space - description of space provided under the title 'The Space'	<br>
description	- summary + space text put together<br>
experiences_offered	- details/description of expereinces offered at the property<br>
neighborhood_overview - brief description of the neighborhood<br>
transit	- transit options avialable close to the property<br>
access - description proovided under the title 'Guest Access'<br>
interaction	- description of host's interaction with the customer<br>
house_rules	- house rules set by the host (listed under policy --> house rules)<br>
picture_url	- url for the picture of property<br>
host_id	- unique ID number for each host<br>
host_url - url where host's profile can be viewed (contains host id)<br>
host_name - name of the host<br>
host_since - date when started hosting on Airbnb<br>
host_location - where the host is located<br>
host_about - brief description provided about themselves by the host<br>
host_response_time - usual response time of the host to customers<br>
host_response_rate - how often the host responds (in %)<br>
host_is_superhost - whether host is a superhost (true/false)<br>
host_thumbnail_url - url to host's thumbnail<br>
host_picture_url - url to host's picture<br>
host_neighbourhood - host's neighborhood<br>
host_listings_count	- number of listings the host has<br>
host_total_listings_count - total listings of the host's properties<br>
host_verifications - forms of verifications of host available<br>
host_has_profile_pic - whether the host has a profile picture<br>
host_identity_verified - whether the host's identity has been verified<br>
street - ???	<br>
neighbourhood - ???	<br>
neighbourhood_cleansed - ???<br>
neighbourhood_group_cleansed - ???<br>
city - city of the property<br>
state - state of the property<br>
zipcode	- zipcode of the property<br>
market - ???<br>
smart_location - ???<br>
country_code - country code (US)	<br>
country	- country name<br>
latitude - lattitude of the property	<br>
longitude - longitude of the property<br>
is_location_exact - whether the location provided is exact (true/false)<br>
property_type - type of property<br>
room_type - entire apartment / private room / shared room<br>
accommodates - how many people can be accomodated<br>
bathrooms - number of bathrooms<br>
bedrooms - number of bedrooms<br>
beds - number of beds<br>
bed_type - type of bed available<br>
amenities - available amenities	<br>
price - price of airbnb per night<br>
security_deposit - amount of security deposit required<br>
cleaning_fee - cleaning fee<br>
guests_included	- how many guests are included in the cost<br>
extra_people - cost of adding an extra person<br>
minimum_nights - minimum number of nights the customer has to book<br>
maximum_nights - maximum number of nights the customer can book<br>
calendar_updated - when was the calender on the listing last updated<br>	
has_availability - whether there is availability (true/false)<br>
availability_30	- number of nights available in the next 30 days (from date when it was scraped 12/6/18)<br>
availability_60	- number of nights available in the next 60 days (from date when it was scraped 12/6/18)<br>
availability_90	- number of nights available in the next 90 days (from date when it was scraped 12/6/18)<br>
availability_365 - number of nights available in the next 365 days (from date when it was scraped 12/6/18)<br>
calendar_last_scraped - date when the calender was last scraped<br>
number_of_reviews - number of reviews for the listing<br>
first_review - date of first review	<br>
last_review	- date of last review<br>
review_scores_rating - overall rating (out of 100)	<br>
review_scores_accuracy	- rating for accuracy (out of 10)<br>
review_scores_cleanliness - rating for cleanliness (out of 10)<br>
review_scores_checkin - rating for checkin (out of 10)<br>
review_scores_communication	- rating for communication (out of 10)<br>
review_scores_location - rating for location (out of 10)<br>
review_scores_value	- rating for value (out of 10)<br>
requires_license - whether license is required (true/false)<br>
instant_bookable - whether it is instantly bookable (true/false)<br>
is_business_travel_ready - whether it is business travel ready (true/false)	<br>
cancellation_policy	- description of the cancellation policy<br>
require_guest_profile_picture - whether guest profice picture is required (true/false)	<br>
require_guest_phone_verification - whether guest phone verification is required (true/false)<br>
calculated_host_listings_count - number of listings by the host<br>
reviews_per_month - average number of reviews per month

First we need to check if there are any listings which are being repeated. We can do this by checking for duplicate values in ID and name columns

In [10]:
data['id'].value_counts().sort_values(ascending=False).head()

21233664    1
23744773    1
27503425    1
17702916    1
20539354    1
Name: id, dtype: int64

In [11]:
data['name'].value_counts().sort_values(ascending=False).head()

New york Multi-unit building    31
Hillside Hotel                  18
New York Apartment              17
Home away from home             13
Brooklyn Apartment              13
Name: name, dtype: int64

This shows us that each of the listing has an unique ID associated with it. There are listings with the same name but different IDs. These can still be considered as seperate listings. One reason for several listings with same name could be that they are: <br>
> accomodation in a hotel/apartment building - i.e. same place but several listings <br>
generic names (such are private room/ suite)>

But, we do not have enough evidence to remove any of the rows due to redundancy at this point

### Next, we can check which columns have redundant data, i.e. which columns have just one value being repeated in each row. These columns are redundant and won't be useful features for modelling.

In [12]:
redundant_data_cols=data.apply(lambda x: x.nunique()).sort_values().head(7)

In [13]:
redundant_data_cols=list(redundant_data_cols.index)
redundant_data_cols

['has_availability',
 'scrape_id',
 'last_scraped',
 'is_business_travel_ready',
 'experiences_offered',
 'requires_license',
 'calendar_last_scraped']

We found the list of columns which have redundant values and we can safely drop them from the dataframe.

In [14]:
#dropping the redundant columns
data.drop(redundant_data_cols, axis=1, inplace=True)

data.shape

(49056, 79)

Seven additional columns were dropped